go/runtime/client/host.go

```diff
@@ -0,0 +1,179 @@
+	go h.worker()
+	return nil
+}
+
+// Stop asks the client host to stop.
+func (h *clientHost) Stop() {
+	close(h.stopCh)
+}
+
+// Quit returns the channel which will signal when the client host has quit.
+func (h *clientHost) Quit() <-chan struct{} {
+	return h.quitCh
+}
+
+// Implements runtimeRegistry.RuntimeHostHandlerFactory.
+func (h *clientHost) GetRuntime() runtimeRegistry.Runtime {
+	return h.runtime
+}
+
+// Implements runtimeRegistry.RuntimeHostHandlerFactory.
+func (h *clientHost) NewRuntimeHostHandler() protocol.Handler {
+	return h
+}
+
+// Implements runtimeRegistry.RuntimeHostHandlerFactory.
+func (h *clientHost) NewNotifier(ctx context.Context, host host.Runtime) protocol.Notifier {
+	return nil
+}
+
+// Implements protocol.Handler.
+func (h *clientHost) Handle(ctx context.Context, body *protocol.Body) (*protocol.Body, error) {
+	// Storage.
+	if body.HostStorageSyncRequest != nil {
+		rq := body.HostStorageSyncRequest
+		span, sctx := opentracing.StartSpanFromContext(ctx, "storage.Sync")
+		defer span.Finish()
+
+		var rs syncer.ReadSyncer
+		switch rq.Endpoint {
+		case protocol.HostStorageEndpointRuntime:
+			// Runtime storage.
+			rs = h.runtime.Storage()
+		case protocol.HostStorageEndpointConsensus:
+			// Consensus state storage.
+			rs = h.consensus.State()
+		default:
+			return nil, errEndpointNotSupported
+		}
+
+		var rsp *storage.ProofResponse
+		var err error
+		switch {
+		case rq.SyncGet != nil:
+			rsp, err = rs.SyncGet(sctx, rq.SyncGet)
+		case rq.SyncGetPrefixes != nil:
+			rsp, err = rs.SyncGetPrefixes(sctx, rq.SyncGetPrefixes)
+		case rq.SyncIterate != nil:
+			rsp, err = rs.SyncIterate(sctx, rq.SyncIterate)
+		default:
+			return nil, errMethodNotSupported
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		return &protocol.Body{HostStorageSyncResponse: &protocol.HostStorageSyncResponse{ProofResponse: rsp}}, nil
+	}
+	// Local storage.
+	if body.HostLocalStorageGetRequest != nil {
+		value, err := h.runtime.LocalStorage().Get(body.HostLocalStorageGetRequest.Key)
+		if err != nil {
+			return nil, err
+		}
+		return &protocol.Body{HostLocalStorageGetResponse: &protocol.HostLocalStorageGetResponse{Value: value}}, nil
+	}
+	if body.HostLocalStorageSetRequest != nil {
+		if err := h.runtime.LocalStorage().Set(body.HostLocalStorageSetRequest.Key, body.HostLocalStorageSetRequest.Value); err != nil {
+			return nil, err
+		}
+		return &protocol.Body{HostLocalStorageSetResponse: &protocol.Empty{}}, nil
+	}
+
+	return nil, errMethodNotSupported
+}
+
+func (h *clientHost) worker() {
+	defer close(h.quitCh)
+
+	// Wait for consensus sync.
+	select {
+	case <-h.consensus.Synced():
+	case <-h.stopCh:
+		return
+	}
+
+	// Start hosted runtime provisioning.
+	hrt, _, err := h.ProvisionHostedRuntime(context.Background())
+	switch {
+	case err == nil:
+	case errors.Is(err, runtimeRegistry.ErrRuntimeHostNotConfigured):
+		// Runtime host not configured, do nothing and wait to stop.
+		<-h.stopCh
```

Comment:
how do we get in this situation?

Reply:
So currently we create a new "client host" for each runtime regardless of whether the hosted runtime is actually configured (e.g., if the client node has specified the path to the runtime binary in the config). But thinking about it, this doesn't make much sense and we should just do the check early and not create this instance at all if the hosted runtime is not configured.

Reply:
That new idea sounds good. What was the nature of "each runtime"? Every single runtime registered on the network, or is there some other list somewhere?

Reply:
Oh, no so each runtime that is enabled in the configuration. So you configure two things:
* One is whether the client node will track the given runtime (e.g., blocks emitted by the runtime) and be able to submit transactions. To do that you just need to configure a runtime ID in the client node's configuration.
* The other is whether the client node can also host a runtime (e.g., has the runtime binary available locally). To do that you need to have the runtime binary available locally and be able to run it (host it).

Reply:
Changed.